build.gradle.kts

```diff
@@ -0,0 +1,385 @@
+import kotlinx.html.js.*
+import org.jetbrains.kotlin.gradle.targets.js.nodejs.NodeJsSetupTask
+import org.apache.tools.ant.taskdefs.condition.Os
+
+/**
+ * This build script supports following parameters:
+ * -Prelease      - activates "release" profile (uploading to Bintray "kotlin/kotlinx.html" without publication and
+ *                  publication to npmjs "kotlinx-html").
+ * -Pbranch-build - activates "branch-build" profile (publication to Bintray "kotlin/kotlin-dev" and to npmjs
+ *                  "@kotlinx-branch-build/kotlinx-html").
+ * -Pmaster-build - activates "master-build" profile (publication to OSS Sonatype snapshot repository and to npmjs
+ *                  "@kotlinx-master-build/kotlinx-html").
+ * -PversionTag   - works together with "branch-build" profile and overrides "-SNAPSHOT" suffix of the version.
+ */
+plugins {
+    kotlin("multiplatform") version "1.3.61"
+    id("maven-publish")
+}
+
+group = "org.jetbrains.kotlinx"
+version = "0.6.13-SNAPSHOT"
+
+/**
+ * If "release" profile is used the "-SNAPSHOT" suffix of the version is removed.
+ */
+if (hasProperty("release")) {
+    val versionString = version as String
+    if (versionString.endsWith("-SNAPSHOT")) {
+        version = versionString.replace("-SNAPSHOT", "")
+    }
+}
+
+/**
+ * Handler of "versionTag" property.
+ * Required to support Bintray and NPM repositories that doesn't support "-SNAPSHOT" versions. To build and publish
+ * artifacts with specific version run "./gradlew -PversionTag=my-tag" and the final version will be "0.6.13-my-tag".
+ */
+if (hasProperty("versionTag")) {
+    val versionString = version as String
+    val versionTag = properties["versionTag"]
+    if (versionString.endsWith("-SNAPSHOT")) {
+        version = versionString.replace("-SNAPSHOT", "-$versionTag")
+        logger.lifecycle("Project will be built with version '$version'.")
+    } else {
+        error("Could not apply 'versionTag' together with non-snapshot version.")
+    }
+}
+
+publishing {
+    publications {
+        repositories {
+            when {
+                hasProperty("release") -> {
+                    maven {
+                        url = uri("https://api.bintray.com/maven/kotlin/kotlinx.html/kotlinx.html;publish=0")
+                        credentials {
```

Comment:
consider extracting `url, username, password`

Reply:
There is almost nothing but this `url`, `username` and `password`. If we extract them we'll just get a lot of unstructured parameters at the top of the script and pointless structures below, won't we?

BTW, I haven't seen such extraction in Ktor either in the coroutines project.

Reply:
Got the idea, but the result is even longer and less readable. Would prefer to keep this version.